File: FloatItPhone/src/main/java/net/geekstools/floatshort/PRO/Util/UI/Splash/FloatingSplash.java
/*
 * Copyright © 2020 By Geeks Empire.
 *
 * Created by Elias Fazel on 1/13/20 7:13 AM
 * Last modified 1/13/20 6:54 AM
 *
 * Licensed Under MIT License.
 * https://opensource.org/licenses/MIT
 */

package net.geekstools.floatshort.PRO.Util.UI.Splash;

import android.app.Service;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.os.IBinder;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.RelativeLayout;

import net.geekstools.floatshort.PRO.R;
import net.geekstools.floatshort.PRO.Util.Functions.FunctionsClass;
import net.geekstools.floatshort.PRO.Util.UI.CustomIconManager.LoadCustomIcons;
import net.geekstools.imageview.customshapes.ShapesImage;

public class FloatingSplash extends Service {

    public static int xPostionRemoval, yPositionRemoval, HWRemoval;
    FunctionsClass functionsClass;
    WindowManager windowManager;
    WindowManager.LayoutParams layoutParams;
    ViewGroup viewGroup;
    RelativeLayout splashView;
    ShapesImage shapesImage;
    int xPosition, yPosition, appColor;
    int statusBarHeight, navBarHeight;
    String packageName, className, inFrontPackageName;
    LoadCustomIcons loadCustomIcons;

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, final int flags, final int startId) {
        windowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
        LayoutInflater layoutInflater = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        navBarHeight = 0;
        int resourceIdNav = getResources().getIdentifier("navigation_bar_height", "dimen", "android");
        if (resourceIdNav > 0) {
            navBarHeight = getResources().getDimensionPixelSize(resourceIdNav);
        }
        statusBarHeight = 0;
        int resourceIdStatus = getResources().getIdentifier("status_bar_height", "dimen", "android");
        if (resourceIdStatus > 0) {
            statusBarHeight = getResources().getDimensionPixelSize(resourceIdStatus);
        }

        packageName = intent.getStringExtra("packageName");
        Drawable appIcon = getDrawable(R.drawable.ic_launcher_balloon);
        try {
            if (intent.hasExtra("className")) {
                className = intent.getStringExtra("className");
                appIcon = functionsClass.shapedAppIcon(getPackageManager().getActivityInfo(new ComponentName(packageName, className), 0)).mutate();
            } else {
                appIcon = functionsClass.loadCustomIcons() ?
                        loadCustomIcons.getDrawableIconForPackage(packageName, functionsClass.shapedAppIcon(packageName).mutate())
                        :
                        functionsClass.shapedAppIcon(packageName).mutate();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        appColor = functionsClass.extractDominantColor(appIcon);
        final int HW = intent.getIntExtra("HW", 0);
        xPosition = intent.getIntExtra("X", 0);
        yPosition = intent.getIntExtra("Y", 0);

        xPostionRemoval = xPosition;
        yPositionRemoval = yPosition + statusBarHeight;
        HWRemoval = HW;

        viewGroup = (ViewGroup) layoutInflater.inflate(R.layout.reveal_splash, null, false);
        splashView = (RelativeLayout) viewGroup.findViewById(R.id.splashView);
        shapesImage = functionsClass.initShapesImage(viewGroup, R.id.shapedIcon);
        RelativeLayout.LayoutParams layoutParamsRelativeLayout = new RelativeLayout.LayoutParams(
                HW,
                HW
        );
        shapesImage.setX(xPosition);
        shapesImage.setY(yPosition + statusBarHeight);
        shapesImage.setLayoutParams(layoutParamsRelativeLayout);
        shapesImage.setImageDrawable(appIcon);
        layoutParams = functionsClass.splashRevealParams(
                0,
                0
        );
        windowManager.addView(viewGroup, layoutParams);
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                try {
                    functionsClass.circularRevealSplashScreen(
                            splashView,
                            shapesImage,
                            xPosition,
                            yPosition + statusBarHeight,
                            appColor,
                            packageName,
                            className,
                            true
                    );
                } catch (Exception e) {
                    e.printStackTrace();
                    if (className != null) {
                        functionsClass.appsLaunchPad(packageName, className);
                    } else {
                        functionsClass.appsLaunchPad(packageName);
                    }
                }
            }
        }, 100);

        viewGroup.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent homeScreen = new Intent(Intent.ACTION_MAIN);
                homeScreen.addCategory(Intent.CATEGORY_HOME);
                homeScreen.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                startActivity(homeScreen);
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        stopSelf();
                    }
                }, 113);
            }
        });

        return START_NOT_STICKY;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        functionsClass = new FunctionsClass(getApplicationContext());

        if (functionsClass.loadCustomIcons()) {
            loadCustomIcons = new LoadCustomIcons(getApplicationContext(), functionsClass.customIconPackageName());
        }
    }

    @Override
    public void onDestroy() {
        try {
            layoutParams.windowAnimations = android.R.style.Animation_Dialog;
            windowManager.updateViewLayout(viewGroup, layoutParams);

            if (functionsClass.UsageStatsEnabled()) {
                sendBroadcast(new Intent("FloatingSplashRemoval"));
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        windowManager.removeViewImmediate(viewGroup);
                    }
                }, 888);
            } else {
                functionsClass.circularRevealSplashScreen(
                        splashView,
                        shapesImage,
                        xPosition,
                        yPosition + statusBarHeight,
                        appColor,
                        packageName,
                        className,
                        false
                );
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        windowManager.removeViewImmediate(viewGroup);
                    }
                }, 133);
            }
        } catch (Exception e) {
            e.printStackTrace();
            windowManager.removeViewImmediate(viewGroup);
        }
    }
}

File: FloatItPhone/src/main/java/net/geekstools/floatshort/PRO/Utils/UI/Splash/FloatingSplash.kt
/*
 * Copyright © 2020 By Geeks Empire.
 *
 * Created by Elias Fazel
 * Last modified 8/29/20 3:58 AM
 *
 * Licensed Under MIT License.
 * https://opensource.org/licenses/MIT
 */

package net.geekstools.floatshort.PRO.Utils.UI.Splash

import android.app.Service
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.IBinder
import android.view.LayoutInflater
import android.view.View
import android.view.WindowManager
import android.widget.RelativeLayout
import net.geekstools.floatshort.PRO.R
import net.geekstools.floatshort.PRO.Utils.Functions.FunctionsClassLegacy
import net.geekstools.floatshort.PRO.Utils.UI.CustomIconManager.LoadCustomIcons
import net.geekstools.floatshort.PRO.databinding.RevealSplashBinding
import net.geekstools.imageview.customshapes.ShapesImage


class FloatingSplash : Service() {

    private val functionsClassLegacy: FunctionsClassLegacy by lazy {
        FunctionsClassLegacy(applicationContext)
    }

    private val windowManager: WindowManager by lazy {
        getSystemService(Context.WINDOW_SERVICE) as WindowManager
    }
    private val layoutParams: WindowManager.LayoutParams by lazy {
        functionsClassLegacy.splashRevealParams(
                0,
                0
        )
    }

    private val layoutInflater: LayoutInflater by lazy {
        getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
    }


    lateinit var shapedIcon: ShapesImage


    companion object {
        var xPositionRemoval: Int = 0
        var yPositionRemoval:Int = 0

        var HeightWidthRemoval:Int = 0
    }

    var xPosition = 0
    var yPosition:Int = 0

    var appIconColor:Int = 0

    var statusBarHeight = 0
    var navBarHeight:Int = 0

    private lateinit var appPackageName: String
    private var appClassName: String? = null

    private val loadCustomIcons: LoadCustomIcons by lazy {
        LoadCustomIcons(applicationContext, functionsClassLegacy.customIconPackageName());
    }

    private lateinit var revealSplashBinding: RevealSplashBinding

    val floatingSplashRemoval: FloatingSplashRemoval by lazy {
        FloatingSplashRemoval(applicationContext)
    }

    override fun onBind(intent: Intent?): IBinder? {

        return null
    }

    override fun onStartCommand(initialIntent: Intent?, flags: Int, startId: Int): Int {

        initialIntent?.let { intent ->


            navBarHeight = 0
            val resourceIdNav = resources.getIdentifier("navigation_bar_height", "dimen", "android")
            if (resourceIdNav > 0) {
                navBarHeight = resources.getDimensionPixelSize(resourceIdNav)
            }
            statusBarHeight = 0
            val resourceIdStatus = resources.getIdentifier("status_bar_height", "dimen", "android")
            if (resourceIdStatus > 0) {
                statusBarHeight = resources.getDimensionPixelSize(resourceIdStatus)
            }

            appPackageName = intent.getStringExtra("packageName")
            var appIcon = getDrawable(R.drawable.ic_launcher_balloon)

            if (intent.hasExtra("className")) {

                appClassName = intent.getStringExtra("className")

                appIcon = if (functionsClassLegacy.customIconsEnable()) loadCustomIcons.getDrawableIconForPackage(packageManager.getActivityInfo(ComponentName(appPackageName, appClassName!!), 0), functionsClassLegacy.shapedAppIcon(appPackageName).mutate()) else functionsClassLegacy.shapedAppIcon(appPackageName).mutate()

            } else {

                appIcon = if (functionsClassLegacy.customIconsEnable()) loadCustomIcons.getDrawableIconForPackage(appPackageName, functionsClassLegacy.shapedAppIcon(appPackageName).mutate()) else functionsClassLegacy.shapedAppIcon(appPackageName).mutate()

            }

            appIconColor = functionsClassLegacy.extractDominantColor(appIcon)
            val HW = intent.getIntExtra("HW", 0)
            val xPosition = intent.getIntExtra("X", 0)
            val yPosition = intent.getIntExtra("Y", 0)

            FloatingSplash.xPositionRemoval = xPosition
            FloatingSplash.yPositionRemoval = yPosition + statusBarHeight
            FloatingSplash.HeightWidthRemoval = HW

            revealSplashBinding = RevealSplashBinding.inflate(layoutInflater)
            shapedIcon = functionsClassLegacy.initShapesImage(revealSplashBinding.shapedIcon)
            val layoutParamsRelativeLayout = RelativeLayout.LayoutParams(
                    HW,
                    HW
            )
            shapedIcon.x = xPosition.toFloat()
            shapedIcon.y = yPosition + statusBarHeight.toFloat()
            shapedIcon.layoutParams = layoutParamsRelativeLayout
            shapedIcon.setImageDrawable(appIcon)

            windowManager.addView(revealSplashBinding.root, layoutParams)

            try {
                val confirmButtonLayoutParams = RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.MATCH_PARENT, RelativeLayout.LayoutParams.MATCH_PARENT)
                floatingSplashRemoval.layoutParams = confirmButtonLayoutParams

                floatingSplashRemoval.visibility = View.INVISIBLE
                revealSplashBinding.rootView.addView(floatingSplashRemoval)
            } catch (e: Exception) {
                e.printStackTrace()
            }

            Handler().postDelayed({

                try {

                    functionsClassLegacy.circularRevealSplashScreen(
                            revealSplashBinding.splashView,
                            shapedIcon,
                            xPosition,
                            yPosition + statusBarHeight,
                            appIconColor,
                            appPackageName,
                            appClassName,
                            true
                    )

                } catch (e: Exception) {
                    e.printStackTrace()

                    if (appClassName != null) {
                        functionsClassLegacy.appsLaunchPad(appPackageName, appClassName)
                    } else {
                        functionsClassLegacy.appsLaunchPad(appPackageName)
                    }
                }

            }, 159)

            revealSplashBinding.root.setOnClickListener {

                val homeScreen = Intent(Intent.ACTION_MAIN)
                homeScreen.addCategory(Intent.CATEGORY_HOME)
                homeScreen.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                startActivity(homeScreen)

                Handler().postDelayed({
                    stopSelf()
                }, 159)

            }

        }

        return Service.START_NOT_STICKY
    }

    override fun onCreate() {
        super.onCreate()

        if (functionsClassLegacy.customIconsEnable()) {

            loadCustomIcons.load()

        }

    }

    override fun onDestroy() {

        try {

            layoutParams.windowAnimations = android.R.style.Animation_Dialog
            windowManager.updateViewLayout(revealSplashBinding.root, layoutParams)

            if (functionsClassLegacy.UsageStatsEnabled()) {

                floatingSplashRemoval.floatingSplashRemoval()

                Handler().postDelayed({
                    windowManager.removeViewImmediate(revealSplashBinding.root)
                }, 888)

            } else {

                functionsClassLegacy.circularRevealSplashScreen(
                        revealSplashBinding.splashView,
                        shapedIcon,
                        xPosition,
                        yPosition + statusBarHeight,
                        appIconColor,
                        appPackageName,
                        appClassName,
                        false
                )
                Handler().postDelayed({

                    windowManager.removeViewImmediate(revealSplashBinding.root)

                }, 133)

            }

        } catch (e: Exception) {
            e.printStackTrace()

            windowManager.removeViewImmediate(revealSplashBinding.root)

        }
    }
}